libbpf-tools/syscount.c

```diff
@@ -0,0 +1,455 @@
+		return -1;
+	}
+	if (ret)
+		*ret = val;
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+
+	return vfprintf(stderr, format, args);
+}
+
+static int bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+static int compar_count(const void *dx, const void *dy)
+{
+	__u64 x = ((struct data_ext_t *) dx)->count;
+	__u64 y = ((struct data_ext_t *) dy)->count;
+	return x > y ? -1 : !(x == y);
+}
+
+static int compar_latency(const void *dx, const void *dy)
+{
+	__u64 x = ((struct data_ext_t *) dx)->total_ns;
+	__u64 y = ((struct data_ext_t *) dy)->total_ns;
+	return x > y ? -1 : !(x == y);
+}
+
+static const char *comm(int comm_map_fd, pid_t pid, char *buf, size_t size)
+{
+	int err;
+
+	err = bpf_map_lookup_elem(comm_map_fd, &pid, buf);
+	if (err) {
+		if (env.verbose) {
+			warn("can't find comm for pid %d: %s",
+			     pid, strerror(-err));
+		}
+		snprintf(buf, size, "[pid %d]", pid);
+	}
+	return buf;
+}
+
+static const char *agg_col(__u32 key, char *buf, size_t size, int comm_map_fd)
+{
+	char comm_buf[COMM_NAME_SIZE];
+
+	if (env.process) {
+		snprintf(buf, size, "%-6u %-15s",
+			 key, comm(comm_map_fd, key,
+				   comm_buf, sizeof(comm_buf)));
+	} else {
+		syscall_name(key, buf, size);
+	}
+	return buf;
+}
+
+static const char *agg_colname(void)
+{
+	return (env.process) ? "PID    COMM" : "SYSCALL";
+}
+
+static const char *time_colname(void)
+{
+	return (env.milliseconds) ? "TIME (ms)" : "TIME (us)";
+}
+
+static void print_latency_header(void)
+{
+	printf("%-22s %8s %16s\n", agg_colname(), "COUNT", time_colname());
+}
+
+static void print_count_header(void)
+{
+	printf("%-22s %8s\n", agg_colname(), "COUNT");
+}
+
+static void print_latency(struct data_ext_t *vals, size_t count, int comm_map_fd)
+{
+	double div = env.milliseconds ? 1000000.0 : 1000.0;
+	char buf[2 * COMM_NAME_SIZE];
+
+	print_latency_header();
+	for (int i = 0; i < count && i < env.top; i++)
+		printf("%-22s %8llu %16.3lf\n",
+		       agg_col(vals[i].key, buf, sizeof(buf), comm_map_fd),
+		       vals[i].count, vals[i].total_ns / div);
+	printf("\n");
+}
+
+static void print_count(struct data_ext_t *vals, size_t count, int comm_map_fd)
+{
+	char buf[2 * COMM_NAME_SIZE];
+
+	print_count_header();
+	for (int i = 0; i < count && i < env.top; i++)
+		printf("%-22s %8llu\n",
+		       agg_col(vals[i].key, buf, sizeof(buf), comm_map_fd),
+		       vals[i].count);
+	printf("\n");
+}
+
+static void print_timestamp()
+{
+	time_t now = time(NULL);
+	struct tm tm;
+
+	if (localtime_r(&now, &tm))
+		printf("[%02d:%02d:%02d]\n", tm.tm_hour, tm.tm_min, tm.tm_sec);
+	else
+		warn("localtime_r: %s", strerror(errno));
+}
+
+static bool read_vals(int fd, struct data_ext_t *vals, __u32 *count)
+{
+	__u32 keys[MAX_ENTRIES] = { -1 };
```

Comment:
that's quite a big on-the-stack array which is initialized every single time, make it static maybe?

Reply:
Ok, thanks, there is actually no need to init it at all, removed.